sbi/inference/fmpe/fmpe_base.py

```diff
@@ -0,0 +1,317 @@
+        self._report_convergence_at_end(self.epoch, stop_after_epochs, max_num_epochs)
+
+        # Update summary.
+        self._summary["epochs_trained"].append(self.epoch)
+        # TODO: rename to loss once renaming is done in base class.
+        self._summary["best_validation_log_prob"].append(self._best_val_log_prob)
+
+        # Update tensorboard and summary dict.
+        self._summarize(round_=self._round)
+
+        # Update description for progress bar.
+        if show_train_summary:
+            print(self._describe_round(self._round, self._summary))
+
+        # Avoid keeping the gradients in the resulting network, which can
+        # cause memory leakage when benchmarking.
+        self._neural_net.zero_grad(set_to_none=True)
+
+        return deepcopy(self._neural_net)
+
+    def build_posterior(
+        self,
+        density_estimator: Optional[ConditionalDensityEstimator] = None,
+        prior: Optional[Distribution] = None,
+        sample_with: str = "direct",
+        direct_sampling_parameters: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ) -> DirectPosterior:
+        """Build the posterior distribution.
+
+        Args:
+            density_estimator: Density estimator for the posterior.
+            prior: Prior distribution.
+            sample_with: Sampling method.
+            direct_sampling_parameters: kwargs for DirectPosterior.
+
+        Returns:
+            DirectPosterior: Posterior distribution.
+        """
+        if sample_with != "direct":
+            raise NotImplementedError(
+                "Currently, only direct sampling is supported for FMPE."
+            )
+
+        if prior is None:
+            assert self._prior is not None, (
+                "You did not pass a prior. You have to pass the prior either at "
+                "initialization `inference = SNPE(prior)` or to "
+                "`.build_posterior(prior=prior)`."
+            )
+            prior = self._prior
+        else:
+            utils.check_prior(prior)
+
+        if density_estimator is None:
+            posterior_estimator = self._neural_net
+            # If internal net is used device is defined.
+            device = self._device
+        else:
+            posterior_estimator = density_estimator
+            # Otherwise, infer it from the device of the net parameters.
+            device = next(density_estimator.parameters()).device.type
+
+        self._posterior = DirectPosterior(
+            posterior_estimator=posterior_estimator,  # type: ignore
+            prior=prior,
+            device=device,
+            **direct_sampling_parameters or {},
+        )
+
+        return deepcopy(self._posterior)
+
+    def append_simulations(
+        self,
+        theta: torch.Tensor,
+        x: torch.Tensor,
+        proposal: Optional[DirectPosterior] = None,
+        exclude_invalid_x: Optional[bool] = None,
+        data_device: Optional[str] = None,
+    ) -> NeuralInference:
+        if (
```

Comment:
what if the user does not pass a proposal but does multi-round? For SNPE we raise an error for this (and tell them to use `force_first_round_loss=True` if they _really_ want to do that), will the same happen for FMPE?